libraries/protocol/custom_authorities/restriction_predicate.cpp

```diff
@@ -0,0 +1,189 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <graphene/protocol/restriction_predicate.hpp>
+
+#include "restriction_predicate.hxx"
+#include "sliced_lists.hxx"
+
+namespace graphene { namespace protocol {
+
+restriction_predicate_function get_restriction_predicate(vector<restriction> rs, operation::tag_type op_type) {
+   auto f = typelist::runtime::dispatch(operation::list(), op_type, [&rs](auto t) -> restriction_predicate_function {
+      using Op = typename decltype(t)::type;
+      if (typelist::contains<operation_list_1::list, Op>())
+         return get_restriction_predicate_list_1(typelist::index_of<operation_list_1::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_2::list, Op>())
+         return get_restriction_predicate_list_2(typelist::index_of<operation_list_2::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_3::list, Op>())
+         return get_restriction_predicate_list_3(typelist::index_of<operation_list_3::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_4::list, Op>())
+         return get_restriction_predicate_list_4(typelist::index_of<operation_list_4::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_5::list, Op>())
+         return get_restriction_predicate_list_5(typelist::index_of<operation_list_5::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_6::list, Op>())
+         return get_restriction_predicate_list_6(typelist::index_of<operation_list_6::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_7::list, Op>())
+         return get_restriction_predicate_list_7(typelist::index_of<operation_list_7::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_8::list, Op>())
+         return get_restriction_predicate_list_8(typelist::index_of<operation_list_8::list, Op>(), std::move(rs));
+      if (typelist::contains<operation_list_9::list, Op>())
+         return get_restriction_predicate_list_9(typelist::index_of<operation_list_9::list, Op>(), std::move(rs));
```

Comment:
Personally I don't like the long list of `if`s, but perhaps it's just me. We can improve it in the future anyway.

For reference, in my first (partial) implementation I tried to do it via `switch-case`, see https://github.com/bitshares/bitshares-core/pull/1362/files#diff-e9075b54e174b51fa26c8db6f535b8f2R507.

Reply:
Yeah, this list annoys me too. It's there to break up the build and keep too many template instantiations from happening in one compilation unit. Unfortunately, it's annoying to get rid of because it's so arbitrary. You could replace it with some kind of mapping or jump table of list-type-to-callable but it's not clear to me that it would be an improvement.